app/src/test/java/org/vss/AbstractKVStoreIntegrationTest.java

```diff
@@ -163,6 +179,177 @@ void getShouldReturnCorrectValueWhenKeyExists() {
+    }
+
+    ListKeyVersionsResponse previousPage = null;
+    List<KeyValue> allKeyVersions = new ArrayList<>();
+    String keyPrefix = "1";
+
+    while (previousPage == null || !previousPage.getKeyVersionsList().isEmpty()) {
+      ListKeyVersionsResponse currentPage;
+
+      if (previousPage == null) {
+        currentPage = list(null, pageSize, keyPrefix);
+      } else {
+        String nextPageToken = previousPage.getNextPageToken();
+        currentPage = list(nextPageToken, pageSize, keyPrefix);
+      }
+
+      allKeyVersions.addAll(currentPage.getKeyVersionsList());
+
+      // Each page.size() is less than or equal to pageSize in request.
+      assertThat(currentPage.getKeyVersionsList().size(), lessThanOrEqualTo(pageSize));
+      previousPage = currentPage;
+    }
+
+    Set<String> uniqueKeys =
+        allKeyVersions.stream().map(KeyValue::getKey).collect(Collectors.toSet());
+
+    // Returns keys only with provided keyPrefix
+    assertThat(uniqueKeys.size(), is(11));
+    assertThat(uniqueKeys,
+        is(Set.of("1k", "10k", "11k", "12k", "13k", "14k", "15k", "16k", "17k", "18k", "19k")));
+  }
+
+  @Test
+  void listShouldReturnZeroGlobalVersionWhenGlobalVersioningNotEnabled() {
+    int totalKvObjects = 1000;
+    for (int i = 0; i < totalKvObjects; i++) {
+      putObjects(null, List.of(kv("k" + i, "k1v1", 0)));
+    }
+
+    ListKeyVersionsResponse previousPage = null;
+    List<KeyValue> allKeyVersions = new ArrayList<>();
+
+    while (previousPage == null || !previousPage.getKeyVersionsList().isEmpty()) {
+      ListKeyVersionsResponse currentPage;
+
+      if (previousPage == null) {
+        currentPage = list(null, null, null);
+
+        // Ensure first page returns global version as ZERO
+        assertThat(currentPage.getGlobalVersion(), is(0L));
+      } else {
+        String nextPageToken = previousPage.getNextPageToken();
+        currentPage = list(nextPageToken, null, null);
+
+        // Ensure pages after first page do not contain global version.
+        assertThat(currentPage.hasGlobalVersion(), is(false));
+      }
+
+      allKeyVersions.addAll(currentPage.getKeyVersionsList());
+      previousPage = currentPage;
+    }
+    // Returns complete view.
+    Set<String> uniqueKeys = allKeyVersions.stream().map(KeyValue::getKey).distinct()
+        .collect(Collectors.toSet());
+    assertThat(uniqueKeys.size(), is(totalKvObjects));
+
+    // Ensure that we don't return "vss_global_version" as part of keys.
+    assertFalse(uniqueKeys.contains(KVStore.GLOBAL_VERSION_KEY));
+  }
+
+  @Test
+  void listShouldLimitMaxPageSize() {
+
+    int totalKvObjects = 10000;
+
+    // Each implementation is free to choose its own max_page_size but there should be a reasonable max
+    // keeping scalability and performance in mind.
+    // Revisit this test case if some implementation wants to support higher page size.
+    int vssArbitraryPageSizeMax = 3000;
```

Comment:
Not sure I follow the reason for this test. Isn't there already a lower constant in the code?

Reply:
MAX_PAGE_SIZE is implementation specific. 
Yes, there is a lower page_size(100) currently configured for postgres implementation but its not going to be same for all impl.

Since these tests lie in AbstractKVStore, we just test that there is some limit to page size from implementation and the behaviour in case cllient requests without page size or with page_size too large.